drivers/i2c/i2c_tca9546a.c

```
@@ -0,0 +1,213 @@
+	const struct device *dev)
+{
+	return dev->config;
+}
+
+static inline struct tca9546a_root_data *
+get_root_data_from_channel(const struct device *dev)
+{
+	return get_channel_config(dev)->root->data;
+}
+
+static inline const struct tca9546a_root_config *
+get_root_config_from_channel(const struct device *dev)
+{
+	return get_channel_config(dev)->root->config;
+}
+
+static int tca9546a_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct tca9546a_root_config *cfg = get_root_config_from_channel(
+		dev);
+
+	return i2c_configure(cfg->bus, dev_config);
+}
+
+static int tca9546a_set_selection(const struct device *dev, uint8_t sel)
+{
+	int res = 0;
+	struct tca9546a_root_data *data = get_root_data(dev);
+	const struct tca9546a_root_config *cfg = get_root_config(dev);
+
+	if (data->select_reg != sel) {
+		res = i2c_write(cfg->bus, &sel, 1, cfg->slave_addr);
+		if (res == 0) {
+			data->select_reg = sel;
+		}
+	}
+	return res;
+}
+
+static int tca9546a_transfer(const struct device *dev,
+			     struct i2c_msg *msgs,
+			     uint8_t num_msgs,
+			     uint16_t addr)
+{
+	struct tca9546a_root_data *data = get_root_data_from_channel(dev);
+	const struct tca9546a_root_config *config = get_root_config_from_channel(
+		dev);
+	const struct tca9546a_channel_config *down_cfg = get_channel_config(dev);
+	int res;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+
+	if (res == 0) {
+		res = tca9546a_set_selection(down_cfg->root, down_cfg->addr);
+		if (res != 0) {
+			goto end_trans;
+		}
+
+		res = i2c_transfer(config->bus, msgs, num_msgs, addr);
+	}
+end_trans:
+	k_mutex_unlock(&data->lock);
+	return res;
+}
+
+static int tca9546_root_init(const struct device *dev)
+{
+	struct tca9546a_root_data *i2c_tca9546a = get_root_data(dev);
+	const struct tca9546a_root_config *config = get_root_config(dev);
+
+	if (!device_is_ready(config->bus)) {
+		LOG_ERR("I2C bus %s not ready", config->bus->name);
+		return -1;
```

maybe better to return an errno error